lib/analyzer/image-extractor.ts

```diff
@@ -0,0 +1,198 @@
+import { createReadStream } from "fs";
+import * as minimatch from "minimatch";
+import { basename } from "path";
+import { PassThrough, Readable, Stream } from "stream";
+import { extract, Extract } from "tar-stream";
+import { streamToBuffer, streamToString } from "../stream-utils";
+
+type ProcessedFileContent = string | Buffer;
+
+type ExtractCallback = (data: Buffer | Stream) => Promise<ProcessedFileContent>;
+type ExtractProducts = Record<string, ProcessedFileContent>;
+
+interface ExtractProductsByFilename {
+  [fileName: string]: ExtractProducts;
+}
+
+export interface ExtractAction {
+  actionName: string; // name, should be unique, for this action
+  fileNamePattern: string; // path pattern to look for
+  callback?: ExtractCallback; // convert Buffer into a string or Buffer
+}
+
+/**
+ * Extract key files from the specified TAR stream.
+ * @param layerTarStream image layer as a Readable TAR stream
+ * @param extractActions array of pattern, callbacks pairs
+ * @returns extracted file products
+ */
+async function extractImageLayer(
+  layerTarStream: Readable,
+  extractActions: ExtractAction[],
+): Promise<ExtractProductsByFilename> {
+  return new Promise((resolve, reject) => {
+    const result: ExtractProductsByFilename = {};
+    const tarExtractor: Extract = extract();
+
+    tarExtractor.on("entry", async (headers, stream, next) => {
+      if (headers.type === "file") {
+        const absoluteFileName = `/${headers.name}`;
+        const processedResult = await extractTarLayerAndProcess(
+          absoluteFileName,
+          stream,
+          extractActions,
+        );
+        if (processedResult !== undefined) {
+          result[absoluteFileName] = processedResult;
+        }
+      }
+
+      stream.resume(); // auto drain the stream
+      next(); // ready for next entry
+    });
+
+    tarExtractor.on("finish", () => {
+      // all layer level entries read
+      resolve(result);
+    });
+
+    tarExtractor.on("error", reject);
+
+    layerTarStream.pipe(tarExtractor);
+  });
+}
+
+async function extractTarLayerAndProcess(
+  fileName: string,
+  fileStream: PassThrough,
+  extractActions: ExtractAction[],
+): Promise<Record<string, ProcessedFileContent> | undefined> {
+  const matchedActions = extractActions.filter((action) =>
+    minimatch(fileName, action.fileNamePattern, { dot: true }),
+  );
+
+  if (matchedActions.length > 0) {
+    const buffer = await streamToBuffer(fileStream);
+    return await applyCallbacks(matchedActions, buffer);
```

Comment:
I am not sure we should do this. This means that every file we find must be loaded into memory! These files can be quite big and this is not necessary for all operations. Yes a stream can only be read once but I don't think it is a good reason to load all files into memory.